apps/glusterfs/pendingop.go

```diff
@@ -0,0 +1,87 @@
+//
+// Copyright (c) 2018 The heketi Authors
+//
+// This file is licensed to you under your choice of the GNU Lesser
+// General Public License, version 3 or any later version (LGPLv3 or
+// later), or the GNU General Public License, version 2 (GPLv2), in all
+// cases as published by the Free Software Foundation.
+//
+
+package glusterfs
+
+import (
+	"fmt"
+)
+
+// The pendingop.go file defines the basic structures needed to track
+// life-cycle of database entries w/in Heketi. There are generally two
+// levels of objects which we track: the pending operation a higher-level
+// operation that should roughly correspond to an action in the Heketi API,
+// and lower level change tracking where individual object creates, deletes,
+// and modifications are logged. Note that objects such as bricks and
+// volumes are still managed in their own buckets. The pending operation
+// metadata tracks the IDs of these objects and these objects have references
+// back to their associated pending operations (via IDs).
+
+// PendingOperationType identifies what kind of high-level operation a
+// PendingOperation will be.
+type PendingOperationType int
+
+const (
+	OperationUnknown PendingOperationType = iota
+	OperationCreateVolume
+	OperationDeleteVolume
+	OperationExpandVolume
+	OperationCreateBlockVolume
+	OperationDeleteBlockVolume
+)
+
+// PendingChangeType identifies what kind of lower-level new item or change
+// is being made to the system as part of a higher-level pending operation.
+type PendingChangeType int
+
+const (
+	OpUnknown PendingChangeType = iota
+	OpAddBrick
+	OpAddVolume
+	OpDeleteBrick
+	OpDeleteVolume
+	OpExpandVolume
+	OpAddBlockVolume
+	OpDeleteBlockVolume
+)
+
+// PendingOperationAction tracks individual changes to entries within the
+// heketi db. It consists of a required change type and (heketi uuid) id,
+// as well as an optional delta object for extra metadata.
+type PendingOperationAction struct {
+	Change PendingChangeType
+	Id     string
+	Delta  interface{}
```

Comment:
wondering if we need parent operation ID here. Just for debugging, logging and Db export cases.

Reply:
I get you, but I don't think its needed because these are only ever found "nestled" within the PendingOperationEntry. If we started embedding or using them outside the pending operation I might agree, but I'd skip it for now.

Reply:
Agreed 👍